essencium-backend/src/main/java/de/frachtwerk/essencium/backend/controller/access/AccessAwareJsonFilter.java

```diff
@@ -24,45 +24,79 @@
 import com.fasterxml.jackson.databind.ser.PropertyWriter;
 import com.fasterxml.jackson.databind.ser.impl.SimpleBeanPropertyFilter;
 import de.frachtwerk.essencium.backend.model.AbstractBaseUser;
+import de.frachtwerk.essencium.backend.model.ApiTokenUser;
 import de.frachtwerk.essencium.backend.model.Ownable;
 import de.frachtwerk.essencium.backend.model.Role;
 import java.io.Serializable;
 import java.util.Arrays;
 import java.util.stream.Stream;
 import lombok.AllArgsConstructor;
 import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
 
 @AllArgsConstructor
-public class AccessAwareJsonFilter<USER extends AbstractBaseUser<ID>, ID extends Serializable>
+public class AccessAwareJsonFilter<
+        U extends UserDetails, USER extends AbstractBaseUser<ID>, ID extends Serializable>
     extends SimpleBeanPropertyFilter {
-  private USER principal;
+  private U userDetails;
 
   @Override
   public void serializeAsField(
       Object pojo, JsonGenerator jsonGenerator, SerializerProvider provider, PropertyWriter writer)
       throws Exception {
-    JsonAllowFor ann = writer.getMember().getAnnotation(JsonAllowFor.class);
-    if (include(writer)
-        && (ann == null
-            || Arrays.stream(ann.roles())
-                .anyMatch(s -> principal.getRoles().stream().map(Role::getName).anyMatch(s::equals))
-            || Stream.of(ann.rights())
-                .anyMatch(
-                    r ->
-                        principal.getAuthorities().stream()
-                            .map(GrantedAuthority::getAuthority)
-                            .anyMatch(r::equals))
-            || (ann.allowForOwner() && isOwner(pojo)))) {
+
+    if (userDetails instanceof ApiTokenUser principal) {
+      JsonAllowFor ann = writer.getMember().getAnnotation(JsonAllowFor.class);
+      if (isAllowedToAccess(principal, ann)) {
+        serializeField(pojo, jsonGenerator, provider, writer);
+      }
+    }
+
+    if (userDetails instanceof AbstractBaseUser<?> principal) {
```

Comment:
Assuming you don't expect any userDetails other than ApiTokenUser and AbstractBaseUser then I would propose to rewrite this to if + else if + else {throw} instead of silently doing nothing